chapters/missing-data.qmd

```diff
@@ -0,0 +1,598 @@
+
+```{r}
+#| label: missing-data-summaries
+#| cache: true
+#| message: false
+#| warning: false
+
+#Note:  Remove quality predictors. These are subjective and related to Sale Price
+ames_original <-
+  ames_original %>%
+  dplyr::select(!contains("Qual")) %>%
+  mutate(Neighborhood = factor(Neighborhood)) %>%
+  arrange(Neighborhood, Order)
+
+ames_predictors <-
+  ames_original %>%
+  dplyr::select(-Order, -PID, -SalePrice)
+
+ames_pct_missing <- 
+  100 * sum(is.na(ames_predictors)) / (nrow(ames_predictors) * ncol(ames_predictors))
+  
+ames_missing <-
+  ames_predictors %>%
+  summarize_all(~ sum(is.na(.)))
+
+ames_num_missing_predictors <- length(which(ames_missing > 0))
+
+ames_missing_long <-
+  ames_predictors %>%
+  mutate(RowNum = row_number()) %>%
+  mutate(across(everything(), as.character)) %>%
+  pivot_longer(cols = -RowNum, names_to = "Predictor", values_to = "value") %>%
+  mutate(Missing = ifelse(is.na(value), "Yes", "No")) %>%
+  mutate(Missing = factor(Missing, levels = c("Yes", "No"))) %>%
+  dplyr::select(-value)
+
+#Remove predictors with more than 20% missing data
+ames_missing_summary <-
+  ames_missing %>%
+  pivot_longer(cols = everything(), names_to = "Predictors", values_to = "value") %>%
+  mutate(`Percent Missing` = 100*value/nrow(ames))
+
+ames_pred_gt_20 <-
+  ames_missing_summary %>%
+  dplyr::filter(`Percent Missing` > 20) %>%
+  dplyr::select(Predictors) %>%
+  pull()
+
+ames_predictors_subset <-
+  ames_predictors %>%
+  dplyr::select(-all_of(ames_pred_gt_20))
+
+ames_pct_missing_subset <- 
+  100 * sum(is.na(ames_predictors_subset)) / 
+  (nrow(ames_predictors_subset) * ncol(ames_predictors_subset))
+
+#Compute the percent missing values within each sample
+binary_fcn <- function(x) ifelse(x == "Yes", 1, 0)
+
+ames_samples <-
+  ames_predictors_subset %>%
+  mutate(RowNum = row_number()) %>%
+  mutate(across(everything(), as.character)) %>%
+  pivot_longer(cols = -RowNum, names_to = "Predictor", values_to = "value") %>%
+  mutate(Missing = ifelse(is.na(value), "Yes", "No")) %>%
+  mutate(Missing = factor(Missing, levels = c("Yes", "No"))) %>% 
+  dplyr::select(-value) %>%
+  pivot_wider(id_cols = "Predictor", names_from = "RowNum", values_from = "Missing") %>%
+  dplyr::select(-Predictor) %>%
+  mutate(across(everything(), binary_fcn)) %>%
+  summarize_all( ~ sum(.) / ncol(ames_predictors_subset) * 100)
+  
+max_missing_ames_samples <- max(ames_samples)
+
+# ------------------------------------------------------------------------------
+
+hotels_predictors <-
+  hotel_rates_all %>%
+  dplyr::select(-avg_price_per_room)
+
+hotels_pct_missing <- 
+  100 * sum(is.na(hotels_predictors)) / 
+  (nrow(hotels_predictors) * ncol(hotels_predictors))
+
+hotels_missing <-
+  hotels_predictors %>%
+  summarize_all(~ sum(is.na(.))) %>%
+  data.frame()
+
+hotels_num_missing_predictors <- length(which(hotels_missing > 0))
+
+hotels_pct_missing_agent <- 100 * length(which(is.na(hotel_rates_all$agent))) / nrow(hotel_rates_all)
```

Comment:
Hmm. My chapter on categorical predictors encodes these as unknown. Maybe they should be actually missing values so that effect encodings work without estimating their level.

Reply:
I didn't recall that from the categorical chapter.  I assumed that an agent wasn't used, which was why the entries were missing.

Reply:
#48